src/Compilers/CSharp/Portable/Parser/SlidingTextWindow.cs

```diff
@@ -34,207 +45,165 @@ internal sealed class SlidingTextWindow : IDisposable
+        /// in the source text. <c>CharacterWindow.Count</c> represents how large the chunk is.  Characters
+        /// <c>[0, CharacterWindow.Count)</c> are valid characters within the window, and represents
+        /// the chunk <c>[CharacterWindowStartPositionInText, CharacterWindowEndPositionInText)</c> in <see cref="Text"/>.
+        /// </summary>
+        private int _characterWindowStartPositionInText;
+
+        private bool _disposed;
+
+        private readonly StringTable _strings;
 
         public SlidingTextWindow(SourceText text)
         {
-            _text = text;
-            _basis = 0;
-            _offset = 0;
+            this.Text = text;
             _textEnd = text.Length;
             _strings = StringTable.GetInstance();
-            _characterWindow = s_windowPool.Allocate();
-            _lexemeStart = 0;
+            _characterWindow = new(s_windowPool.Allocate());
+
+            // Read the first chunk of the file into the character window.
+            this.ReadChunkAt(0);
         }
 
         public void Dispose()
         {
-            if (_characterWindow != null)
-            {
-                s_windowPool.Free(_characterWindow);
-                _characterWindow = null!;
-                _strings.Free();
-            }
-        }
+            Debug.Assert(!_disposed);
+            _disposed = true;
 
-        public SourceText Text => _text;
+            s_windowPool.Free(_characterWindow.Array!);
+            _strings.Free();
+        }
 
         /// <summary>
-        /// The current absolute position in the text file.
+        /// Reads a chunk of characters from the underlying <see cref="Text"/> at the given position and places them
+        /// at the start of the character window.  The character windows length will be set to the number of characters
+        /// read.
+        /// <para/>
+        /// Note: this does NOT set _positionInText.  We are just reading the characters into the window. The actual 
+        /// position in the text is unchanged by this, and callers should set that if necessary.
         /// </summary>
-        public int Position
+        /// <param name="position"></param>
+        private void ReadChunkAt(int position)
         {
-            get
-            {
-                return _basis + _offset;
-            }
+            position = Math.Min(position, _textEnd);
+
+            var amountToRead = Math.Min(_textEnd - position, DefaultWindowLength);
+            this.Text.CopyTo(position, _characterWindow.Array!, 0, amountToRead);
+            _characterWindowStartPositionInText = position;
+            _characterWindow = new(_characterWindow.Array!, 0, amountToRead);
         }
 
         /// <summary>
-        /// The current offset inside the window (relative to the window start).
+        /// The current absolute position in the text file.
         /// </summary>
-        public int Offset
```

Comment:
importantly, nearly all exposed internal state of this type is now gone.  External code was neeeding to understand this and keep thigns in sync.  We now expose with much simpler primitives (like ReadOnlySpans) to ensure callers don't need math, or understand the internal config of htis type.